packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts

```diff
@@ -0,0 +1,289 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { CollectionReference } from '@google-cloud/firestore';
+import {
+  AppStoreServerAPI,
+  decodeRenewalInfo,
+  decodeTransaction,
+} from 'app-store-server-api';
+import Container from 'typedi';
+
+import { AppConfig, AuthLogger } from '../../../types';
+import { AppleIAP } from './apple-iap';
+import {
+  APPLE_APP_STORE_FORM_OF_PAYMENT,
+  mergePurchaseWithFirestorePurchaseRecord,
+  SubscriptionPurchase,
+} from './subscription-purchase';
+import {
+  PurchaseQueryError,
+  PurchaseUpdateError,
+  TransactionType,
+} from './types';
+
+/**
+ * This class wraps Firestore API calls for Apple IAP purchases. It
+ * will handle related CRUD operations.
+ * This is using the V2 App Store Server API implementation:
+ * https://developer.apple.com/documentation/appstoreserverapi
+ */
+export class PurchaseManager {
+  private appStoreConfig: any;
+  private log: AuthLogger;
+
+  /*
+   * This class is intended to be initialized by the library.
+   * Library consumer should not initialize this class themselves.
+   */
+  constructor(
+    private purchasesDbRef: CollectionReference,
+    private appStoreDeveloperApiClients: {
+      [key: string]: AppStoreServerAPI;
+    }
+  ) {
+    this.log = Container.get(AuthLogger);
+    this.appStoreConfig = Container.get(AppConfig).subscriptions.appStore;
+  }
+
+  /*
+   * Query a subscription purchase by its bundle ID and original transaction ID.
+   * The method queries the V2 App Store Server API to get the latest status of the purchase,
+   * then merge it with purchase ownership info stored in the library's managed Firestore database,
+   * then returns the merged information as a SubscriptionPurchase to its caller.
+   * A brand new subscription will not yet have a defined userId;
+   * that's handled by registerToUserAccount.
+   */
+  public async querySubscriptionPurchase(
+    bundleId: string,
+    originalTransactionId: string
+  ) {
+    // STEP 1. Query App Store Server API to get the subscription status
+    let apiResponse;
+    let subscriptionStatus;
+    let transactionInfo;
+    let renewalInfo;
+    try {
+      apiResponse = await this.appStoreDeveloperApiClients[
+        bundleId
+      ].getSubscriptionStatuses(originalTransactionId);
+      // Find the latest transaction for the subscription
+      const item = apiResponse.data[0].lastTransactions.find(
+        (item) => item.originalTransactionId === originalTransactionId
+      );
+      subscriptionStatus = item.status;
+      transactionInfo = await decodeTransaction(item.signedTransactionInfo);
+      renewalInfo = await decodeRenewalInfo(item.signedRenewalInfo);
```

Comment:
This is kind of a bummer, I looked at the implementation, its identical for each of these which means a bit of repeat work. Hopefully its not re-fetching certs as well but I'm not entirely sure. Putting a FIXME or TODO above this pointing out possible inefficiency and optimization here would be good.

Reply:
I'll file a follow-up to look into this more and link it here.